src/dex/uniswap-v3/uniswap-v3.ts

```diff
@@ -0,0 +1,824 @@
+    dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network] || {},
+    readonly routerIface = new Interface(UniswapV3RouterABI),
+    readonly quoterIface = new Interface(UniswapV3QuoterABI),
+    protected config = UniswapV3Config[dexKey][network],
+    protected poolsToPreload = PoolsToPreload[dexKey][network] || [],
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.uniswapMulti = new this.dexHelper.web3Provider.eth.Contract(
+      UniswapV3MultiABI as AbiItem[],
+      this.config.uniswapMulticall,
+    );
+    // To receive revert reasons
+    this.dexHelper.web3Provider.eth.handleRevert = false;
+
+    // Normalise once all config addresses and use across all scenarios
+    this.config = this._toLowerForAllConfigAddresses();
+  }
+
+  get supportedFees() {
+    return this.config.supportedFees;
+  }
+
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return this.adapters[side] ? this.adapters[side] : null;
+  }
+
+  getPoolIdentifier(srcAddress: Address, destAddress: Address, fee: bigint) {
+    const tokenAddresses = this._sortTokens(srcAddress, destAddress).join('_');
+    return `${this.dexKey}_${tokenAddresses}_${fee}`;
+  }
+
+  async initializePricing(blockNumber: number) {
+    // This is only for testing, because cold pool fetching is goes out of
+    // FETCH_POOL_INDENTIFIER_TIMEOUT range
+    await Promise.all(
+      this.poolsToPreload.map(async pool =>
+        Promise.all(
+          this.config.supportedFees.map(async fee =>
+            this.getPool(pool.token0, pool.token1, fee, blockNumber),
+          ),
+        ),
+      ),
+    );
+  }
+
+  async getPool(
+    srcAddress: Address,
+    destAddress: Address,
+    fee: bigint,
+    blockNumber: number,
+  ): Promise<UniswapV3EventPool | null> {
+    let pool =
+      this.eventPools[this.getPoolIdentifier(srcAddress, destAddress, fee)];
+    if (pool === undefined) {
+      const [token0, token1] = this._sortTokens(srcAddress, destAddress);
+
+      const key = `${token0}_${token1}_${fee}`.toLowerCase();
+      await this.dexHelper.cache.hset(
+        this.dexmapKey,
+        key,
+        JSON.stringify({
+          token0,
+          token1,
+          fee: fee.toString(),
+        }),
+      );
+
+      this.logger.info(`starting to listen to new pool: ${key}`);
+      pool = new UniswapV3EventPool(
+        this.dexHelper,
+        this.dexKey,
+        this.config.stateMulticall,
+        this.config.factory,
+        fee,
+        token0,
+        token1,
+        this.logger,
+      );
+
+      try {
+        await pool.initialize(blockNumber, {
+          initCallback: (state: DeepReadonly<PoolState>) => {
+            //really hacky, we need to push poolAddress so that we subscribeToLogs in StatefulEventSubscriber
+            pool!.addressesSubscribed[0] = state.pool;
+            pool!.poolAddress = state.pool;
```

Comment:
May you please, use type narrowing instead of exclamation mark? I really think that using exclamation marks should be only when it is really needed. Because you are bypassing TS type checking system and make it useless:
```
          if (pool !== null) {
            pool.addressesSubscribed.push(state.pool);
            pool.poolAddress = state.pool;
          } else {
            // Logging for impossible case
          }
```

I saw that you assigned previously new UniswapV3EventPool to `pool` and I don't know too why TS is not picking up automatically that `pool` is not `null` anymore. This change will be helpful if we do some refactor and someone removes accidentally that assignment, then we will catch in logs, that it was broken. But with exclamation mark it will go unnoticed.

Reply:
Ts is not picking up and there is no chance pool is null

Reply:
There is no chance unless someone moves this line for whatever reason somewhere outside:

```
      pool = new UniswapV3EventPool(
        this.dexHelper,
        this.dexKey,
        this.config.stateMulticall,
        this.config.factory,
        fee,
        token0,
        token1,
        this.logger,
      );
```